### PR TITLE
JWT auth for realtime

### PR DESCRIPTION
**Pull request by aninternetof**
This adds JWT auth to realtime. The auth is handled by a new custom middleware class.

You can test it out with the [in-progress branch of rovercode-ui that adds the websocket](https://github.com/rovercode/rovercode-ui/tree/Add-rover-api-websocket).

### CODE REVIEW

File: realtime/consumers.py [hunk 1]
```diff
@@ -11,10 +12,29 @@ class RoverConsumer(WebsocketConsumer):
...
+                oauth_application__client_id=self.room_name)
+        except Rover.DoesNotExist:
+            self.close()
+            return
+
+        if rover.owner.id != user.id:
+            self.close()
+            return
+
+        print("Realtime: user {} connected to rover {}"
```
[the remark this answers is not in the document]
Comment (cabarnes) [reply realtime/consumers.py]: Yes, it will

File: realtime/consumers.py [hunk 2]
```diff
@@ -11,10 +12,29 @@ class RoverConsumer(WebsocketConsumer):
     room_group_name = None
 
     def connect(self):
-        """Handle connections."""
+        """Handle new connection requests."""
+        user = self.scope.get('user')
+        if not user or user.is_anonymous:
+            self.close()
```
[the remark this answers is not in the document]
Comment (aninternetof) [reply realtime/consumers.py]: Ok, looking at the [WebsocketConsumer implementation](https://github.com/django/channels/blob/master/channels/generic/websocket.py#L83), it looks like the best I can do is return a code. When I tested it, I do see the code, but no other space for a response. On a success, it's a `200 Switched Protocol`, so maybe that kind of request doesn't look for a response body? Or, maybe I could just shove the message in the code, and make it `code="4XX Some message"`?

image

For now, I made it just return the code. Luckily, 401, 403, and 404 seem to cover our three cases here in a meaningful way.

I made this one 401.

File: realtime/consumers.py [hunk 3]
```diff
@@ -11,10 +12,29 @@ class RoverConsumer(WebsocketConsumer):
...
+            return
+
         self.room_name = self.scope['url_route']['kwargs']['room_name']
         self.room_group_name = 'chat_%s' % self.room_name
 
+        try:
+            rover = Rover.objects.get(
+                oauth_application__client_id=self.room_name)
+        except Rover.DoesNotExist:
+            self.close()
```
[the remark this answers is not in the document]
Comment (aninternetof) [reply realtime/consumers.py]: See note above. I made this one 404.

File: realtime/consumers.py [hunk 4]
```diff
@@ -11,10 +12,29 @@ class RoverConsumer(WebsocketConsumer):
...
+            return
+
+        if rover.owner.id != user.id:
+            self.close()
```
[the remark this answers is not in the document]
Comment (aninternetof) [reply realtime/consumers.py]: See note above. I made this one 403.

File: realtime/tests/test_rover_consumer.py [hunk 5]
```diff
@@ -22,3 +45,73 @@
     assert response == message
     # Close
     await communicator.disconnect()
+
+
+@pytest.mark.django_db(transaction=True)
+@pytest.mark.asyncio
+async def test_rover_consumer_no_user():
+    """Test sending a message to the room and having it echo it back."""
```
[the remark this answers is not in the document]
Comment (aninternetof) [reply realtime/tests/test_rover_consumer.py]: Fixed.

File: realtime/consumers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aninternetof) [reply realtime/consumers.py]: Done.

**Comment by aninternetof** [comment]
By some magic, the localhost:800/realtime/roomname debugging endpoints get a jwt when logged in through that rovercode-web log-in screens, so they still work!

**Pull request merged**